### PR TITLE
Move functionality of reading Emails from disk

### LINKED ISSUE #192: Disk manipulating methods in ICEmail
The methods `maildirFileName`, `deleteMailFromDisk` and `saveInMailDir` should not be implemented in the ICEmail class.

Actually they should not be implemented anywhere as Disk interaction in Squeak should happen once and then never again. We should plan on changing the mail storage on the disk to a mail storage in the image #212

### LINKED ISSUE #212: Remove FileIO, store data inside image
also fixes #192, #227.

### PR DESCRIPTION
**Pull request by MrHenrik**
Closes #212 #192

### CODE REVIEW

File: packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/readEmailsFor.From.in..st [hunk 1]
```diff
@@ -0,0 +1,7 @@
+read file
+readEmailsFor: aFolder From: aPath in: currentDirectory
```
Comment (Veladus) [inline packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/readEmailsFor.From.in..st]: Casing is inconsistent

File: packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/recursivelyReadFolderStructureWith.and.and.and..st [hunk 2]
```diff
@@ -0,0 +1,17 @@
+read file
```
Comment (Veladus) [inline packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/recursivelyReadFolderStructureWith.and.and.and..st]: I think this should be categorized as private

File: packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/recursivelyReadFolderStructureWith.and.and.and..st [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+recursivelyReadFolderStructureWith: maildirDirectory and: directoryName and: currentFolder and: anEndpoint
```
Comment (Veladus) [inline packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/recursivelyReadFolderStructureWith.and.and.and..st]: I'd argue the distinction between `FileDirectory` and `ICFolder` should be more pronounced

**Review by Veladus — DISMISSED** [review]

**Comment by Veladus** [comment]
TODO:
* Make tests pass
* Make more tests
* refactor `maildirFilename`

File: packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/directoryToFolderDictionary..st [hunk 4]
```diff
@@ -0,0 +1,3 @@
+private
+directoryToFolderDictionary: anObject
```
Comment (shm0sby) [inline packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/directoryToFolderDictionary..st:2]: `anObject` isn't the best name.

File: packages/IMAPClient-Tests.package/ICFileReadWriterTest.class/instance/testReadWrite.st [hunk 5]
```diff
@@ -1,10 +1,4 @@
...
 	
-	| email |
-	
-	self readWriter write: (self class expectedMailObject) in: 'Maildir' as: 'testMail'. 
-	email := (self readWriter get: 'testMail' in: 'Maildir').
-	
-	self assert: (email header) equals: self class expectedMailObject header.
-	self assert: (email body) equals: self class expectedMailObject body
\ No newline at end of file
+	"Todo"
```
Comment (shm0sby) [inline packages/IMAPClient-Tests.package/ICFileReadWriterTest.class/instance/testReadWrite.st]: You don't want to have todos in PRs. 
Why isn't it already part of the PR?

File: packages/IMAPClient-Core.package/ICEmail.class/instance/^equals.st [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+	"Answer whether the receiver and anObject represent the same object."
+
+	self == anObject
+		ifTrue: [ ^ true ].
+	self class = anObject class
+		ifFalse: [ ^ false ].
+	^ self header = anObject header
+		and: [ 
+			self messageID = anObject messageID
+				and: [ self body = anObject body and: [ self folder = anObject folder and: [ self uniqueID = anObject uniqueID ] ] ] ]
```
Comment (shm0sby) [inline packages/IMAPClient-Core.package/ICEmail.class/instance/^equals.st:12]: The alignment of the brackets is a bit weird.

File: packages/IMAPClient-Core.package/ICEmail.class/instance/^equals.st [hunk 7]
```diff
@@ -0,0 +1,12 @@
+comparing
+= anObject
+	"Answer whether the receiver and anObject represent the same object."
```
Comment (shm0sby) [inline packages/IMAPClient-Core.package/ICEmail.class/instance/^equals.st:3]: This comment can be deleted.

File: packages/IMAPClient-Core.package/ICEmail.class/instance/deleteEmailFromDisk.st [hunk 8]
```diff
@@ -2,4 +2,4 @@ action
 deleteEmailFromDisk
 
 	self folder removeMail: self. 
-	FileDirectory default deleteFileNamed: (self maildirFileName, FileDirectory slash, self uniqueID)		
\ No newline at end of file
+	ICFileReadWriter new delete: self
```
Comment (shm0sby) [inline packages/IMAPClient-Core.package/ICEmail.class/instance/deleteEmailFromDisk.st]: Why do we need a new instance here? 
Couldn't that be static?

File: packages/IMAPClient-Core.package/ICEmail.class/instance/deserializeFrom..st [hunk 9]
```diff
@@ -0,0 +1,8 @@
+accessing-header
+deserializeFrom: aJson
+
+	self header: (Dictionary newFrom: (aJson at: 'header' ifAbsent: [JsonObject new]));
```
Comment (shm0sby) [inline packages/IMAPClient-Core.package/ICEmail.class/instance/deserializeFrom..st:4]: Maybe that kind of initialization is worth a mention in the class comment?

File: packages/IMAPClient-Core.package/ICEmail.class/instance/saveInMaildir.st [hunk 10]
```diff
@@ -1,7 +1,4 @@
 accessing
 saveInMaildir
 
-	^ ICFileReadWriter new
-		write: self
-		in: self maildirFileName 
-		as: self uniqueID
\ No newline at end of file
+	^ ICFileReadWriter new write: self
```
Comment (shm0sby) [inline packages/IMAPClient-Core.package/ICEmail.class/instance/saveInMaildir.st]: Same question why this should be a new instance.

File: packages/IMAPClient-Core.package/ICEndPoint.class/instance/login.st [hunk 11]
```diff
@@ -18,6 +18,6 @@ login
 		self fatalErrorOccured: true.
 		^ false].
  
-	self loadEmailsFromMaildir.
+	(ICFileReadWriter newWith: self) readEmailsFromDisk.
```
Comment (shm0sby) [inline packages/IMAPClient-Core.package/ICEndPoint.class/instance/login.st:21]: Is there anything similar for the closing of the IMAP client to write them down?
Or how do they become persistent?

File: packages/IMAPClient-Tests.package/ICEndPointCommandTest.class/instance/testRetrieveFlagsOnly.st [hunk 12]
```diff
@@ -35,9 +34,13 @@ testRetrieveFlagsOnly
 	self assert: mailB flagged equals: true.
 	
 	"Asserts that the flags - in the maildir storage file - were also updated"
-	readWriter := ICFileReadWriter new.
-	mailA := readWriter get: '3' in: path.
-	mailB := readWriter get: '4' in: path.
+	
+	streamA :=	FileStream readOnlyFileNamed: path , FileDirectory slash, '3'.
```
Comment (shm0sby) [inline packages/IMAPClient-Tests.package/ICEndPointCommandTest.class/instance/testRetrieveFlagsOnly.st]: Strange indentation

File: packages/IMAPClient-Tests.package/ICEndPointCommandTest.class/instance/testRetrieveFlagsOnly.st [hunk 13]
```diff
@@ -35,9 +34,13 @@ testRetrieveFlagsOnly
...
+	mailA := ICEmail newFromJsonStream: streamA.
+	streamB := FileStream readOnlyFileNamed: path , FileDirectory slash, '4'.
+	mailB := ICEmail newFromJsonStream: streamB.
+	streamA close.
+	streamB close.
 	
 	self assert: mailA seen equals: false.
```
Comment (shm0sby) [inline packages/IMAPClient-Tests.package/ICEndPointCommandTest.class/instance/testRetrieveFlagsOnly.st:40]: Having one assertion per test is the better style.
Try to combine both conditions.

File: packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/write..st [hunk 14]
```diff
@@ -0,0 +1,8 @@
+write file
+write: anICEmail
+	
+	(FileDirectory default / anICEmail maildirFileName) assureExistence.
```
Comment (shm0sby) [inline packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/write..st]: Can we have a test for that if its not existing? 🤔

File: packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/readEmailsFromDisk.st [hunk 15]
```diff
@@ -0,0 +1,6 @@
+read file
+readEmailsFromDisk
+
+	self rootDirectory
+			assureExistence;
```
Comment (shm0sby) [inline packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/readEmailsFromDisk.st:5]: Test for that would be nice.

**Review by shm0sby — CHANGES_REQUESTED** [review]

File: packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/directoryToFolderDictionary..st (hunk 4 above)
Comment (Veladus) [reply packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/directoryToFolderDictionary..st:2]: This is generated.

File: packages/IMAPClient-Tests.package/ICFileReadWriterTest.class/instance/testReadWrite.st (hunk 5 above)
Comment (Veladus) [reply packages/IMAPClient-Tests.package/ICFileReadWriterTest.class/instance/testReadWrite.st]: Writing this test is a little bigger issue, but we want to merge this into develop to have this in future prs.

File: packages/IMAPClient-Core.package/ICEmail.class/instance/^equals.st (hunk 6 above)
Comment (Veladus) [reply packages/IMAPClient-Core.package/ICEmail.class/instance/^equals.st:12]: This is generated.

File: packages/IMAPClient-Core.package/ICEmail.class/instance/^equals.st (hunk 7 above)
Comment (Veladus) [reply packages/IMAPClient-Core.package/ICEmail.class/instance/^equals.st:3]: this is generated.

File: packages/IMAPClient-Core.package/ICEmail.class/instance/deserializeFrom..st (hunk 9 above)
Comment (Veladus) [reply packages/IMAPClient-Core.package/ICEmail.class/instance/deserializeFrom..st:4]: @JuliaJoch

File: packages/IMAPClient-Core.package/ICEndPoint.class/instance/login.st (hunk 11 above)
Comment (Veladus) [reply packages/IMAPClient-Core.package/ICEndPoint.class/instance/login.st:21]: They get written down when they get fetched.

File: packages/IMAPClient-Tests.package/ICEndPointCommandTest.class/instance/testRetrieveFlagsOnly.st (hunk 13 above)
Comment (Veladus) [reply packages/IMAPClient-Tests.package/ICEndPointCommandTest.class/instance/testRetrieveFlagsOnly.st:40]: We know this. This just conforms the test to the new interface, but I will change this to look better.

File: packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/write..st (hunk 14 above)
Comment (Veladus) [reply packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/write..st]: Writing tests is an issue which we know of, but we want to merge this first.

File: packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/readEmailsFromDisk.st (hunk 15 above)
Comment (Veladus) [reply packages/IMAPClient-Core.package/ICFileReadWriter.class/instance/readEmailsFromDisk.st:5]: I don't get, what you mean. This is taken from the standard library.

File: packages/IMAPClient-Core.package/ICEmail.class/instance/deleteEmailFromDisk.st (hunk 8 above)
Comment (Veladus) [reply packages/IMAPClient-Core.package/ICEmail.class/instance/deleteEmailFromDisk.st]: You mean the Mail should hold a reference to a `ICFileReadWriter`?
Putting this on the class side shouldn't be possible as this need access to the `endpoint maildirFilename` which for the mean time `ICEmail` offers as legacy API, but we will remove this shortly after. Then you will need to pass an `ICEndpoint` to create an `ICFileReadWriter`.

Comment (Veladus) [reply packages/IMAPClient-Core.package/ICEmail.class/instance/deleteEmailFromDisk.st]: Just noticed that closing #192 requires to remove `maildirFilename` so I will change this.

**Comment by JuliaJoch** [comment]
`deleteEmailFromDisk` is still called in `checkTopicality` which leads to a message not understood error when updating the account. Please pull this into an encapsulated method and make sure you have refactored all methods which rely on the deleted method!

**Comment by JuliaJoch** [comment]
While moving emails to different folders you can quickly encounter some inconsistent state. When deleting the Maildir folder locally the state is fixed, but we should make sure that this does not happen in the first place. Maybe we should not call the save and update to Maildir Methods as long as it does not have the correct functionality of Maildir anyways?

File: packages/IMAPClient-Tests.package/ICFileReadWriterTest.class/instance/testReadWrite.st (hunk 5 above)
Comment (MrHenrik) [reply packages/IMAPClient-Tests.package/ICFileReadWriterTest.class/instance/testReadWrite.st]: we have now created the tests

**Review by shm0sby — APPROVED** [review]

**Pull request merged**